### PR TITLE
feat(jit): Add pl.RUNTIME for unspecialized scalars in signature-mode compile()

### LINKED ISSUE #2283: [Feature] Support runtime (unspecialized) scalar parameters in annotation-driven compile()
### Summary

`@pl.jit` kernels that carry a `pl.Scalar` parameter meant to be **runtime** (value supplied at dispatch, not baked into the IR) cannot be compiled via annotation-driven "signature mode" (`compile()` with no tensor args). Today the only escape hatch is positional sample-arg mode, which forces the caller to manufacture throwaway tensors for every tensor parameter just to reach the (trailing) scalar slot.

### Motivation / Use Case

`JITFunction.compile()` has two modes (`pypto/jit/decorator.py`):

- **Signature mode** — `compile()` with no positional args; tensor shape/dtype read from `pl.Tensor[[...], dtype]` annotations, dynamic dims stay extent-independent. No dummy tensors needed.
- **Sample-arg mode** — `compile(*args)`; tensor shape/dtype read from the passed torch tensors.

Mode selection (`_resolve_specialization`, `decorator.py:1948`):

```python
signature_mode = allow_signature_mode and not args and not any(_is_tensor(v) for v in kwargs.values())
```

A `pl.Scalar` annotation carries only a *type* (`pl.Scalar[pl.INT32]`), no value. In signature mode the scalar value is resolved by `_signature_scalar_value`, which **requires a literal `int`/`float`/`bool` and specializes that value into the compiled artifact** (`decorator.py:380`):

```python
if not isinstance(value, (int, float, bool)):
    raise TypeError(... "must be an int/float/bool" ...)
return value
```

So a scalar whose value varies at runtime (e.g. a per-step token count) **cannot use signature mode** — any value supplied gets baked in. The only workaround is sample-arg mode, which is positional: passing one positional arg disables signature mode, and `_bind_args` → `sig.bind` then requires a value for *every* parameter. Because such runtime scalars are typically trailing parameters, the caller must build throwaway (zero-storage meta) tensors for all preceding tensor params purely to land a placeholder in the scalar's positional slot.

Concretely, pypto-serving's DeepSeek executor has to build an entire meta-tensor tuple just to compile two kernels that each carry one runtime scalar:

```python
# pypto_serving/model/deepseek/npu_executor.py:199
def _runtime_scalar_compile_args(jit_fn, runtime_scalar_names):
    # tensors        -> torch.empty(shape, dtype, device="meta")  # zero storage
    # runtime scalars -> ctypes.c_int32()                         # deliberately NOT int/float/bool
```

The `ctypes.c_int32()` only works because `_bind_args` classifies scalars by `isinstance(value, (int, float, bool))` (`decorator.py:1794`), so a ctypes value is *not* captured into `scalar_values` and is left unspecialized. That is an implicit, undocumented contract: the supported path (signature mode) forces specialization, and the unspecialized path only works through this incidental type-filtering in positional mode.

### Proposed API / Behavior

A first-class way to mark a `pl.Scalar` parameter as **runtime / unspecialized** so it can be used in signature-mode `compile()`. Some options:

1. **Explicit runtime-scalar marker** — e.g. `pl.RuntimeScalar[pl.INT32]` (or `pl.Scalar[pl.INT32, runtime=True]`), which `_bind_args_from_signature` leaves out of `scalar_values` (treated like a dynamic dim: value independent, supplied at dispatch) instead of calling `_signature_scalar_value`.
2. **Accept a ctypes/sentinel value in signature mode** — relax `_signature_scalar_value` so a ctypes or sentinel value marks the scalar as unspecialized rather than erroring, mirroring what `_bind_args` already does implicitly.
3. **Keyword scalar binding that does not specialize** — allow `compile(my_scalar=<runtime int32>)` to keep signature mode (tensors from annotations) while leaving the scalar runtime-parametric.

Any of these would let kernels with runtime scalars compile from annotations alone and remove the meta-tensor workaround downstream.

### Alternatives Considered

- **Status quo** (sample-arg mode + meta tensors + ctypes placeholder): works today via the incidental `int/float/bool` filter in `_bind_args`, but is undocumented, couples "runtime-ness" to a type-filtering side effect, and forces every caller with a runtime scalar to build a full zero-storage tensor tuple. This is what pypto-serving currently does.

### Additional Context

- pypto commit: `a47d3004`
- Relevant code (`pypto/jit/decorator.py`): `_resolve_specialization` (`:1948`), `_bind_args_from_signature` (`:1799`), `_signature_scalar_value` (`:380`), `_bind_args` scalar classification (`:1794`).
- The `compile()` docstring already documents the no-arg signature path and notes "Scalar parameters have no value in the signature, so pass them as keyword args" (`decorator.py:2131`) — but that path specializes the value, which is exactly the gap for runtime scalars.
- Downstream workaround: `pypto_serving/model/deepseek/npu_executor.py:199` (`_runtime_scalar_compile_args`).

### PR DESCRIPTION
**Pull request by lyfne123**
Annotation-driven signature mode required every pl.Scalar parameter to carry an
int/float/bool, and specialized that value into the compiled artifact. A scalar
whose value varies per launch therefore could not use signature mode at all: the
only escape hatch was positional sample-arg mode, which forced the caller to
build throwaway meta tensors for every preceding tensor parameter just to reach
the scalar's slot, and relied on _bind_args' int/float/bool filter incidentally
skipping a ctypes placeholder.

The lowering pipeline already supported unspecialized scalars — the specializer
only substitutes names present in scalar_values, so a name it does not know
survives as a real pl.Scalar parameter. What was missing was a way to say so.

pl.RUNTIME is that marker:

    compiled = kernel.compile(num_tokens=pl.RUNTIME)

The scalar is omitted from scalar_values (and hence from the cache key, so one
artifact serves every value) while its dtype is still recorded. It also works as
a signature default. Because the marker is compile-time only, _bind_args now
rejects it by name rather than letting it reach the runtime as a bogus scalar.

Dep-level scalar dtype propagation is fixed to forward the dtype independently
of the value, so a runtime scalar keeps its dtype across the JIT call graph.

Closes #2283

### CODE REVIEW

File: python/pypto/jit/decorator.py [hunk 1]
```diff
@@ -2130,8 +2173,11 @@ def compile(self, *args: Any, **kwargs: Any) -> Any:
         raises). Dynamic dims (``pl.dynamic`` / ``bind_dynamic``) need no value —
         the artifact is extent-independent. Scalar parameters have no value in
         the signature, so pass them as keyword args (or via a signature
-        default). This shares the same cache entry as an equivalent
-        ``compile(*sample_tensors)`` call.
+        default); a literal **specializes** that value into the artifact, while
+        ``pl.RUNTIME`` leaves the parameter **unspecialized** — it stays a real
+        ``pl.Scalar`` parameter supplied at dispatch and, like a dynamic dim,
+        drops out of the cache key. This shares the same cache entry as an
+        equivalent ``compile(*sample_tensors)`` call.
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply python/pypto/jit/decorator.py]: Fixed in f78af52a. The sentence was pre-existing, but inserting the `pl.RUNTIME` paragraph directly before it made it read as if RUNTIME shared that entry too — which `test_runtime_scalar_and_literal_do_not_share_cache` explicitly asserts it does not. Reworded to scope cache sharing to agreeing specialized scalars, and to state that the sample-argument path always specializes the value it is handed and rejects the marker.

File: python/pypto/jit/decorator.py [hunk 2]
```diff
@@ -2147,10 +2193,16 @@ def my_kernel(x, w, out):
...
+            # Its value is then supplied on each dispatch through the compiled
+            # artifact (below), not by calling my_kernel(...) directly — an
+            # eager call re-specializes and compiles a separate artifact.
+            compiled = my_kernel.compile(num_tokens=pl.RUNTIME)
+
             w_dev = worker.alloc_tensor(real_w.shape, real_w.dtype, init=real_w)
             h = worker.register(compiled)
             for batch in stream:
-                h(batch.x, w_dev, batch.out)
+                h(batch.x, w_dev, batch.out, batch.num_tokens)
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply python/pypto/jit/decorator.py]: Fixed in f78af52a. The arity mismatch was mine — I added a `num_tokens` compile/dispatch to a three-parameter example. The example kernel now declares `num_tokens: pl.Scalar[pl.INT32]` alongside fully shaped `pl.Tensor[[...], dtype]` annotations (with `M = pl.dynamic("M")`), so the signature-mode claim in the surrounding prose actually holds, and the sample-argument line passes a literal `128` for the scalar. `real_w` / `stream` stay as illustrative stand-ins for caller-supplied data, consistent with `sample_x` / `batch` elsewhere in the same block.

File: tests/ut/jit/test_jit_compile_extraction.py [hunk 3]
```diff
@@ -374,6 +410,85 @@ def scalar_sig_kernel(
...
+        dispatch call, or a sample-argument compile — must reject it by name
+        instead of letting it reach the runtime as a bogus scalar. The signature
+        default makes this reachable from a plain ``kernel(a, c)`` call."""
+        torch = pytest.importorskip("torch")
+
+        t = torch.zeros(256, 128, dtype=torch.float32)
+        with pytest.raises(TypeError, match="'n' received pl.RUNTIME"):
+            rt_scalar_default_kernel._bind_args((t, t), {})
+        with pytest.raises(TypeError, match="'n' received pl.RUNTIME"):
+            rt_scalar_kernel._bind_args((t, pl.RUNTIME, t), {})
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply tests/ut/jit/test_jit_compile_extraction.py]: Fixed in f78af52a — both sites now use raw escaped patterns: `r"'n' received pl\.RUNTIME"` (473, 475) and `r"pl\.RUNTIME"` (489). Verified with ruff 0.16 `--select RUF043`: no findings remain in the lines this PR touches. Note the repo pins ruff 0.14.8 in `.pre-commit-config.yaml`, which does not flag RUF043, so this was not a CI failure — the escaping is still right on its own merits, since the unescaped `.` made the assertion looser than it reads. The one remaining RUF043 hit in this file (line 389, `match="bare 'pl.Tensor'"`) is pre-existing and outside this PR's scope.

**Review by Hzfengsy — APPROVED** [review]
